### ISSUE TITLE
maxConnection count relies on onClose not throwing an exception

### ISSUE THREAD

**Issue by bbangert** [tavendo/AutobahnPython#501]
maxConnections watches countConnections to determine remaining connections to allow. However, if an exception is thrown in onClose, the countConnections is not decremented. This means an error in onClose can easily starve your system of allowing connections.

The _onClose call from WebSocketProtocol's _connectionLost should either be wrapped in a try/finally to ensure countConnections is decremented, or the count should be decremented before calling the _connectionLost handler here: https://github.com/tavendo/AutobahnPython/blob/master/autobahn/websocket/protocol.py#L2759

Or it should be very very clear to folks implementing this that any exception in onClose could cause this to occur.

**Comment by meejah** [comment]
Yes, the actual call to "user"-overridable code should be guarded with try/except. For example, https://github.com/tavendo/AutobahnPython/blob/master/autobahn/twisted/websocket.py#L175 should do this wrapping (like in the WAMP classes). Same for anything we call from the ``IWebSocketChannel`` interface I guess.

_connectionLost itself shouldn't throw an exception, ever -- but that's a bug in _connectionLost itself then.

Related to this: these methods (onClose, etc) should be documented as to whether they're allowed to be async (return a Deferred/Future) and if so handle that properly in the calling code (e.g. add errbacks).
